aic-emu/GfxHandler.cc

```diff
@@ -0,0 +1,542 @@
+
+GfxHandler::GfxHandler(const char* deviceStr)
+{
+    if (deviceStr)
+        m_device_str = deviceStr;
+    else
+        m_device_str = DEFAULT_GFX_DEVICE;
+}
+
+GfxHandler::~GfxHandler()
+{
+    std::map<uint64_t, boData_t>::iterator it;
+
+    for (it = m_boList.begin(); it != m_boList.end(); it++)
+    {
+        ClearBo(it->first);
+    }
+    m_boList.clear();
+
+    if (m_device_fd > 0)
+    {
+        GfxStatus status = GFX_OK;
+        status = GfxClose();
+        GFX_LOG_STATUS(status != GFX_OK, " GfxClose");
+    }
+}
+
+GfxStatus GfxHandler::RunIoctl(unsigned long request, void* buf)
+{
+   int sts = GFX_OK;
+
+   if (!buf)
+   {
+       std::cout <<std::string(__FUNCTION__) << ": Non-null pointer required for buf" << std::endl;
+       return GFX_FAIL;
+   }
+
+   do
+   {
+     sts = ioctl(m_device_fd, request, buf);
+   }
+   while (sts == -1 && (errno == EINTR || errno == EAGAIN));
+
+   std::string errorString = std::string(__FUNCTION__) + ": Req " + std::to_string(request);
+   GFX_LOG_FAILURE(sts != 0, errorString.c_str());
+
+   return (GfxStatus) sts;
+}
+
+GfxStatus GfxHandler::GetDrmParam(unsigned long request, int* val)
+{
+   if (!val)
+   {
+       std::cout <<std::string(__FUNCTION__) << ": Non-null pointer required for val" << std::endl;
+       return GFX_FAIL;
+   }
+
+   struct drm_i915_getparam gp = { .param = (int)request, .value = val };
+
+   return RunIoctl(DRM_IOCTL_I915_GETPARAM, &gp);
+}
+
+GfxStatus GfxHandler::GfxInit()
+{
+    int sts = GFX_OK;
+    int mmap_gtt_version = -1;
+
+    if(m_device_str.empty())
+    {
+        std::cout << "Error: Empty device string in Init" << std::endl;
+        sts = GFX_FAIL;
+        goto exit_logic;
+    }
+    else
+        std::cout << "Using device " << m_device_str << std::endl;
+
+    //Init device props to 0
+    m_device_props = {.chipsetId = 0, .hasMMAPoffset = 0};
+
+    //Open device
+    m_device_fd = open(m_device_str.c_str(), O_RDWR);
+    GFX_LOG_STATUS((m_device_fd < 0), "open device");
+    if (!GFX_SUCCESS(m_device_fd < 0))
+        goto exit_logic;
+
+    //Get version info
+    struct drm_version drm_ver;
+    memset(&drm_ver, 0, sizeof(drm_ver));
+
+    sts = RunIoctl(DRM_IOCTL_VERSION, &drm_ver);
+
+    GFX_LOG_FAILURE(sts != GFX_OK, "drmIoctl DRM_IOCTL_VERSION first");
+    if (!GFX_SUCCESS(sts))
+        goto exit_logic;
+
+    //Allocate memory and call the ioctl again to actually get the name
+    if (drm_ver.name_len > 0)
+    {
+        drm_ver.name = (char *)malloc(drm_ver.name_len + 1);
+
+        sts = RunIoctl(DRM_IOCTL_VERSION, &drm_ver);
+
+        GFX_LOG_FAILURE(sts != GFX_OK, "drmIoctl DRM_IOCTL_VERSION second");
+
+        if (drm_ver.name_len >=0)
+            drm_ver.name[drm_ver.name_len] = '\0';
+
+        m_device_props.version = (const char *) drm_ver.name;
+
+        if (drm_ver.name)
+            free(drm_ver.name);
+
+        if (!GFX_SUCCESS(sts))
+            goto exit_logic;
+    }
+
```

Comment:
What was the reason to get the name? I guess intent was to check that we are dealing with i915 and abort if we do not...

Reply:
yeah, that was sort of the reason: to run through basic calls to the kernel, and print a start message with SW info (the version name), HW (device chipset) and required cap (in this MMAP offset support) info.
the program aborts if any of these calls fails. we could expand this to abort for other fail conditions, but we can take that up when we add extend this further e.g. with configurability added for local vs system memory support.